spec/core/matchers/toHaveSpyInteractions.js

```diff
@@ -0,0 +1,57 @@
+describe('toHaveSpyInteractions', () => {
+  let spyObj;
+  beforeEach(() => {
+    spyObj = jasmineUnderTest.createSpyObj('NewClass', ['spyA', 'spyB']);
+    spyObj.otherMethod = function() {};
+  });
+
+  it('detects spy interactions', () => {
+    spyObj.spyA();
+    expect(spyObj).toHaveSpyInteractions();
+  });
+
+  it('detects multiple spy interactions', () => {
+    spyObj.spyA();
+    spyObj.spyB();
+    spyObj.spyA();
+    expect(spyObj).toHaveSpyInteractions();
+  });
+
+  it('detects no spy interactions', () => {
+    expect(spyObj).not.toHaveSpyInteractions();
+  });
+
+  it('ignores non-observed spy object interactions', () => {
+    spyObj.otherMethod();
+    expect(spyObj).not.toHaveSpyInteractions();
+  });
+
+  [true, 123, 'string'].forEach(testValue => {
+    it(`throws error if a non-object (${testValue}) is passed`, () => {
+      expect(() => {
+        expect(true).toHaveSpyInteractions();
+      }).toThrowError(Error, /Expected a spy object, but got/);
+    });
+  });
+
+  [['argument'], [false, 0]].forEach(testValue => {
```

Comment:
What's the significance of `[false, 0]`? It seems to me that this could probably be reduced down to a single spec, but if we need both cases I'd rather see something that makes it clear to future maintainers (e.g. me in 6 months) why those specific values were chosen.

Reply:
Please don't use this form of test in Jasmine's suite, especially for this small of a set of possible options. As @sgravrock points out, it is generally hard to understand what is really being tested, and is not in the style of the rest of the Jasmine repository.

Additionally, the `${testValue}` interpolation will only work inside of Node.js and not in the browsers that Jasmine also tests against, leaving test failures very hard to track down.

Reply:
I will simplify `  [['argument'], [false, 0]].forEach(function(testValue) {` and just use one test. The relevance here is simply to have one or multiple arguments, but I do not think the value is large.


@slackersoft 
However, what's the suggestion on this test?

```
  [true, 123, 'string'].forEach(function(testValue) {
    it(`throws error if a non-object (${testValue}) is passed`, function() {
      let spyObj = jasmineUnderTest.getEnv().createSpyObj('NewClass', ['spyA', 'spyB']);

      expect(function() {
        expect(testValue).toHaveSpyInteractions();
      }).toThrowError(Error, /Expected a spy object, but got/);
    });
  });
```

This is trying to make sure that the logic is detecting as many as non-objects as possible. In order to avoid this for loop, I could think of two options:


A) I would need to create one new test for each of the possible types. This would go against DRY principle as we will have 3 tests (potentially more in the future) with exactly the same code, testing exactly the same principle and just changing the input. 

B) I would need to do the loop inside the function. However, if the tests fails, it would be hard to debug which exact variant is failing, which partially defeats the purpose of testing.

What's your suggested approach here?

Reply:
I think aiming for DRY in your tests is usually harmful to the readability and debuggability of your test suite, and as a rule, Jasmine does _NOT_ try to achieve DRY-ness in its own test suite. An array of 3 items, probably doesn't even deserve a loop. If you look at tests for other matchers, you'll see that we tend to have a single spec labeled something like "throws an error if a non-object is passed" and then inside that single test multiple explicit calls to `expect` with the different values.